lib/Sema/ConstraintSystem.cpp

```diff
@@ -2734,34 +2734,50 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
   // overloads of the same declaration.
   ConstraintLocator *commonCalleeLocator = nullptr;
   SmallPtrSet<ValueDecl *, 4> distinctChoices;
-  SmallVector<std::pair<const Solution *, const ConstraintFix *>, 4>
-      viableSolutions;
+  SmallVector<const Solution *, 4> viableSolutions;
+
+  enum class AmbiguityKind {
+    /// There is exactly one fix associated with each candidate.
+    CloseMatch,
+    /// Solution is ambiguous because all candidates had partially matching
+    /// parameter lists.
+    ParameterList,
+    /// General ambiguity failure.
+    General,
+  };
+  auto ambiguityKind = AmbiguityKind::CloseMatch;
 
   bool diagnosable = llvm::all_of(solutions, [&](const Solution &solution) {
     ArrayRef<ConstraintFix *> fixes = solution.Fixes;
 
-    // Currently only support a single fix in a solution,
-    // but ultimately should be able to deal with multiple.
-    if (fixes.size() != 1)
+    if (fixes.empty())
       return false;
 
-    const auto *fix = fixes.front();
-    auto  *locator = fix->getLocator();
+    if (ambiguityKind == AmbiguityKind::CloseMatch && fixes.size() > 1)
```

Comment:
What do you think about this condition:

```c++
if (fixes.size() > 1) {
  ambiguityKind = llvm::all_of(fixes, [](const ConstraintFix *fix) -> bool {
    auto *locator = fix->getLocator();
    return locator->findLast<LocatorPathElt::ApplyArgument>();
  }) 
  ? AmbiguityKind::ParameterList 
  : AmbiguityKind::General;
}
```

I think it's okay to check for the common locator as a separate pass to keep this simple since it's not that performance sensitive.

Reply:
I do think it's better to determine the ambiguity kind separately, thanks! I'll make this change